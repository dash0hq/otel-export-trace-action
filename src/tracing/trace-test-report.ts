import {
  Span,
  TraceAPI,
  Tracer,
  Context,
  SpanStatusCode,
  SpanStatus,
} from "@opentelemetry/api";
import { parse } from "test-results-parser";
import TestCase = require("test-results-parser/src/models/TestCase");
import TestStep = require("test-results-parser/src/models/TestStep");
import TestSuite = require("test-results-parser/src/models/TestSuite");

export type TraceJunitArtifactParams = {
  trace: TraceAPI;
  tracer: Tracer;
  parentContext: Context;
  parentSpan: Span;
  startTime: Date;
  path: string;
  type: string;
};

export function traceTestReportArtifact({
  tracer,
  trace,
  parentSpan,
  startTime,
  parentContext,
  path,
  type,
}: TraceJunitArtifactParams) {
  if (!["junit", "xunit", "ngunit"].includes(type)) {
    throw TypeError(
      `Report tracing only supports junit, xunit, or ngunit. ${type} is not supported`
    );
  }
  const result = parse({ type, files: [path] });
  parentSpan.setAttributes({
    "tests.type": type,
    "tests.total": result.total,
    "tests.name": result.name,
    "tests.passed": result.passed,
    "tests.failed": result.failed,
    "tests.errors": result.errors,
    "tests.skipped": result.skipped,
    "tests.retried": result.retried,
    "tests.status": result.status,
  });

  result.suites.map((testSuite) => {
    traceTestSuite({
      startTime,
      testSuite,
      parentContext,
      parentSpan,
      tracer,
      trace,
    });
  });
  parentSpan.setStatus({
    code: result.status === "PASS" ? SpanStatusCode.OK : SpanStatusCode.ERROR,
  });
}
type TraceTestSuiteParams = {
  testSuite: TestSuite;
  parentContext: Context;
  parentSpan: Span;
  startTime: Date;
  tracer: Tracer;
  trace: TraceAPI;
};

function traceTestSuite({
  testSuite,
  parentContext,
  parentSpan,
  startTime,
  trace,
  tracer,
}: TraceTestSuiteParams) {
  const attributes = {
    "tests.testSuite.duration": testSuite.duration,
    "tests.testSuite.errors": testSuite.errors,
    "tests.testSuite.failed": testSuite.failed,
    "tests.testSuite.name": testSuite.name,
    "tests.testSuite.passed": testSuite.passed,
    "tests.testSuite.skipped": testSuite.skipped,
    "tests.testSuite.status": testSuite.status,
    "tests.testSuite.total": testSuite.total,
  };
  const ctx = trace.setSpan(parentContext, parentSpan);
  const span = tracer.startSpan(testSuite.name, { startTime, attributes }, ctx);
  span.setStatus({
    code:
      testSuite.status === "PASS" ? SpanStatusCode.OK : SpanStatusCode.ERROR,
  });
  let testCaseStartTime = new Date(startTime);
  try {
    testSuite.cases.map((testCase) => {
      traceTestCases({
        testCase,
        startTime: testCaseStartTime,
        parentContext: ctx,
        parentSpan: span,
        trace,
        tracer,
      });
      testCaseStartTime = new Date(testCaseStartTime);
      testCaseStartTime.setMilliseconds(
        testCaseStartTime.getMilliseconds() + testCase.duration * 1000
      );
    });
  } finally {
    const endTime = new Date(startTime);
    endTime.setMilliseconds(
      startTime.getMilliseconds() + testSuite.duration * 1000
    );
    span.end(endTime);
  }
}

type TraceTestCasesParams = {
  testCase: TestCase;
  parentContext: Context;
  parentSpan: Span;
  startTime: Date;
  tracer: Tracer;
  trace: TraceAPI;
};

function traceTestCases({
  testCase,
  parentContext,
  parentSpan,
  startTime,
  trace,
  tracer,
}: TraceTestCasesParams) {
  const attributes = {
    "tests.testCase.duration": testCase.duration,
    "tests.testCase.errors": testCase.errors,
    "tests.testCase.failed": testCase.failed,
    "tests.testCase.failure": testCase.failure,
    "tests.testCase.name": testCase.name,
    "tests.testCase.passed": testCase.passed,
    "tests.testCase.skipped": testCase.skipped,
    "tests.testCase.stack_trace": testCase.stack_trace,
    "tests.testCase.status": testCase.status,
    "tests.testCase.total": testCase.total,
  };
  const ctx = trace.setSpan(parentContext, parentSpan);
  const span = tracer.startSpan(testCase.name, { startTime, attributes }, ctx);
  const spanStatus: SpanStatus = {
    code: testCase.status === "PASS" ? SpanStatusCode.OK : SpanStatusCode.ERROR,
  };
  if (testCase.failure) {
    spanStatus.message = testCase.failure;
  }
  span.setStatus(spanStatus);
  // let testStepStartTime = new Date(startTime);
  try {
    // testCase.steps.map((testStep) => {
    //   traceTestSteps({
    //     testStep,
    //     parentContext: ctx,
    //     parentSpan: span,
    //     startTime: testStepStartTime,
    //     trace,
    //     tracer,
    //   });
    //   testStepStartTime = new Date(testStepStartTime);
    //   testStepStartTime.setMilliseconds(
    //     testStepStartTime.getMilliseconds() + testStep.duration * 1000
    //   );
    // });
  } finally {
    const endTime = new Date(startTime);
    endTime.setMilliseconds(
      endTime.getMilliseconds() + testCase.duration * 1000
    );
    span.end(endTime);
  }
}

// type TraceTestStepsParams = {
//   testStep: TestStep;
//   parentContext: Context;
//   parentSpan: Span;
//   startTime: Date;
//   trace: TraceAPI;
//   tracer: Tracer;
// };

// function traceTestSteps({
//   testStep,
//   parentContext,
//   parentSpan,
//   startTime,
//   trace,
//   tracer,
// }: TraceTestStepsParams) {
//   const attributes = {
//     "tests.testStep.duration": testStep.duration,
//     "tests.testStep.failure": testStep.failure,
//     "tests.testStep.name": testStep.name,
//     "tests.testStep.stack_trace": testStep.stack_trace,
//     "tests.testStep.status": testStep.status,
//   };
//   const ctx = trace.setSpan(parentContext, parentSpan);
//   const span = tracer.startSpan(testStep.name, { startTime, attributes }, ctx);

//   const endTime = new Date(startTime);
//   endTime.setMilliseconds(
//     startTime.getMilliseconds() + testStep.duration * 1000
//   );
//   span.end(endTime);
// }
