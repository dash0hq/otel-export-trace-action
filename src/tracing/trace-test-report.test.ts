import { traceTestReportArtifact } from "./trace-test-report";
import path from "path";
import { context, trace } from "@opentelemetry/api";
import {
  BasicTracerProvider,
  InMemorySpanExporter,
  SimpleSpanProcessor,
} from "@opentelemetry/sdk-trace-base";
import { Resource } from "@opentelemetry/resources";
import { SemanticResourceAttributes } from "@opentelemetry/semantic-conventions";

function idGenerator() {
  let counter = 1;
  return () => {
    counter++;
    return `123456789${counter}`;
  };
}

describe("traceTestReportArtifact", () => {
  const memoryExporter = new InMemorySpanExporter();
  const tracerProvider = new BasicTracerProvider({
    resource: new Resource({
      [SemanticResourceAttributes.SERVICE_NAME]:
        "traceTestReportArtifact - test",
    }),
    idGenerator: {
      generateTraceId: idGenerator(),
      generateSpanId: idGenerator(),
    },
  });
  tracerProvider.addSpanProcessor(new SimpleSpanProcessor(memoryExporter));
  tracerProvider.register();
  const tracer = trace.getTracer("default");

  beforeEach(() => {
    memoryExporter.reset();
  });

  afterEach(() => {
    // clear require cache
    Object.keys(require.cache).forEach((key) => delete require.cache[key]);
    return tracerProvider.shutdown();
  });

  it("test junit spans match snapshot", () => {
    const junitFilePath = path.join(
      "src",
      "tracing",
      "__assets__",
      "{lint-and-format-check}{run tests}{junit}.xml"
    );
    const startTime = new Date("2022-01-22T04:45:30");

    tracer.startActiveSpan(
      "test step span",
      { startTime, root: true },
      (stepSpan) => {
        traceTestReportArtifact({
          trace,
          tracer,
          context,
          stepSpan,
          startTime,
          path: junitFilePath,
          type: "junit",
        });
        stepSpan.end(new Date("2022-01-22T04:45:34"));
      }
    );

    const spans = memoryExporter.getFinishedSpans();
    expect(spans).toMatchSnapshot();
  });
});
