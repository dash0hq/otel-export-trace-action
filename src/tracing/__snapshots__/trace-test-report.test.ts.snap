// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`traceTestReportArtifact test junit spans match snapshot 1`] = `
Array [
  Span {
    "_attributeValueLengthLimit": Infinity,
    "_duration": Array [
      13,
      0,
    ],
    "_ended": true,
    "_spanContext": Object {
      "spanId": "1234567894",
      "traceFlags": 1,
      "traceId": "1234567894",
      "traceState": undefined,
    },
    "_spanLimits": Object {
      "attributeCountLimit": 128,
      "attributeValueLengthLimit": Infinity,
      "eventCountLimit": 128,
      "linkCountLimit": 128,
    },
    "_spanProcessor": MultiSpanProcessor {
      "_spanProcessors": Array [
        SimpleSpanProcessor {
          "_exporter": InMemorySpanExporter {
            "_finishedSpans": [Circular],
            "_stopped": false,
          },
          "_isShutdown": false,
          "_shuttingDownPromise": Promise {},
        },
      ],
    },
    "attributes": Object {
      "tests.testCase.duration": 13,
      "tests.testCase.errors": 0,
      "tests.testCase.failed": 0,
      "tests.testCase.failure": "",
      "tests.testCase.name": "createTracerProvider test service.name resource",
      "tests.testCase.passed": 0,
      "tests.testCase.skipped": 0,
      "tests.testCase.stack_trace": "",
      "tests.testCase.status": "PASS",
      "tests.testCase.total": 0,
    },
    "endTime": Array [
      1642848343,
      0,
    ],
    "events": Array [],
    "instrumentationLibrary": Object {
      "name": "default",
      "version": undefined,
    },
    "kind": 0,
    "links": Array [],
    "name": "createTracerProvider test service.name resource",
    "parentSpanId": undefined,
    "resource": Resource {
      "attributes": Object {
        "service.name": "traceTestReportArtifact - test",
        "telemetry.sdk.language": "nodejs",
        "telemetry.sdk.name": "opentelemetry",
        "telemetry.sdk.version": "1.0.1",
      },
    },
    "startTime": Array [
      1642848330,
      0,
    ],
    "status": Object {
      "code": 0,
    },
  },
  Span {
    "_attributeValueLengthLimit": Infinity,
    "_duration": Array [
      29,
      0,
    ],
    "_ended": true,
    "_spanContext": Object {
      "spanId": "1234567895",
      "traceFlags": 1,
      "traceId": "1234567895",
      "traceState": undefined,
    },
    "_spanLimits": Object {
      "attributeCountLimit": 128,
      "attributeValueLengthLimit": Infinity,
      "eventCountLimit": 128,
      "linkCountLimit": 128,
    },
    "_spanProcessor": MultiSpanProcessor {
      "_spanProcessors": Array [
        SimpleSpanProcessor {
          "_exporter": InMemorySpanExporter {
            "_finishedSpans": [Circular],
            "_stopped": false,
          },
          "_isShutdown": false,
          "_shuttingDownPromise": Promise {},
        },
      ],
    },
    "attributes": Object {
      "tests.testCase.duration": 29,
      "tests.testCase.errors": 0,
      "tests.testCase.failed": 0,
      "tests.testCase.failure": "",
      "tests.testCase.name": "createTracerProvider test service.instance.id resource",
      "tests.testCase.passed": 0,
      "tests.testCase.skipped": 0,
      "tests.testCase.stack_trace": "",
      "tests.testCase.status": "PASS",
      "tests.testCase.total": 0,
    },
    "endTime": Array [
      1642848372,
      0,
    ],
    "events": Array [],
    "instrumentationLibrary": Object {
      "name": "default",
      "version": undefined,
    },
    "kind": 0,
    "links": Array [],
    "name": "createTracerProvider test service.instance.id resource",
    "parentSpanId": undefined,
    "resource": Resource {
      "attributes": Object {
        "service.name": "traceTestReportArtifact - test",
        "telemetry.sdk.language": "nodejs",
        "telemetry.sdk.name": "opentelemetry",
        "telemetry.sdk.version": "1.0.1",
      },
    },
    "startTime": Array [
      1642848343,
      0,
    ],
    "status": Object {
      "code": 0,
    },
  },
  Span {
    "_attributeValueLengthLimit": Infinity,
    "_duration": Array [
      2,
      0,
    ],
    "_ended": true,
    "_spanContext": Object {
      "spanId": "1234567896",
      "traceFlags": 1,
      "traceId": "1234567896",
      "traceState": undefined,
    },
    "_spanLimits": Object {
      "attributeCountLimit": 128,
      "attributeValueLengthLimit": Infinity,
      "eventCountLimit": 128,
      "linkCountLimit": 128,
    },
    "_spanProcessor": MultiSpanProcessor {
      "_spanProcessors": Array [
        SimpleSpanProcessor {
          "_exporter": InMemorySpanExporter {
            "_finishedSpans": [Circular],
            "_stopped": false,
          },
          "_isShutdown": false,
          "_shuttingDownPromise": Promise {},
        },
      ],
    },
    "attributes": Object {
      "tests.testCase.duration": 2,
      "tests.testCase.errors": 0,
      "tests.testCase.failed": 0,
      "tests.testCase.failure": "",
      "tests.testCase.name": "createTracerProvider test service.namespace resource",
      "tests.testCase.passed": 0,
      "tests.testCase.skipped": 0,
      "tests.testCase.stack_trace": "",
      "tests.testCase.status": "PASS",
      "tests.testCase.total": 0,
    },
    "endTime": Array [
      1642848374,
      0,
    ],
    "events": Array [],
    "instrumentationLibrary": Object {
      "name": "default",
      "version": undefined,
    },
    "kind": 0,
    "links": Array [],
    "name": "createTracerProvider test service.namespace resource",
    "parentSpanId": undefined,
    "resource": Resource {
      "attributes": Object {
        "service.name": "traceTestReportArtifact - test",
        "telemetry.sdk.language": "nodejs",
        "telemetry.sdk.name": "opentelemetry",
        "telemetry.sdk.version": "1.0.1",
      },
    },
    "startTime": Array [
      1642848372,
      0,
    ],
    "status": Object {
      "code": 0,
    },
  },
  Span {
    "_attributeValueLengthLimit": Infinity,
    "_duration": Array [
      3,
      0,
    ],
    "_ended": true,
    "_spanContext": Object {
      "spanId": "1234567897",
      "traceFlags": 1,
      "traceId": "1234567897",
      "traceState": undefined,
    },
    "_spanLimits": Object {
      "attributeCountLimit": 128,
      "attributeValueLengthLimit": Infinity,
      "eventCountLimit": 128,
      "linkCountLimit": 128,
    },
    "_spanProcessor": MultiSpanProcessor {
      "_spanProcessors": Array [
        SimpleSpanProcessor {
          "_exporter": InMemorySpanExporter {
            "_finishedSpans": [Circular],
            "_stopped": false,
          },
          "_isShutdown": false,
          "_shuttingDownPromise": Promise {},
        },
      ],
    },
    "attributes": Object {
      "tests.testCase.duration": 3,
      "tests.testCase.errors": 0,
      "tests.testCase.failed": 0,
      "tests.testCase.failure": "",
      "tests.testCase.name": "createTracerProvider test active span processor",
      "tests.testCase.passed": 0,
      "tests.testCase.skipped": 0,
      "tests.testCase.stack_trace": "",
      "tests.testCase.status": "PASS",
      "tests.testCase.total": 0,
    },
    "endTime": Array [
      1642848377,
      0,
    ],
    "events": Array [],
    "instrumentationLibrary": Object {
      "name": "default",
      "version": undefined,
    },
    "kind": 0,
    "links": Array [],
    "name": "createTracerProvider test active span processor",
    "parentSpanId": undefined,
    "resource": Resource {
      "attributes": Object {
        "service.name": "traceTestReportArtifact - test",
        "telemetry.sdk.language": "nodejs",
        "telemetry.sdk.name": "opentelemetry",
        "telemetry.sdk.version": "1.0.1",
      },
    },
    "startTime": Array [
      1642848374,
      0,
    ],
    "status": Object {
      "code": 0,
    },
  },
  Span {
    "_attributeValueLengthLimit": Infinity,
    "_duration": Array [
      3581,
      0,
    ],
    "_ended": true,
    "_spanContext": Object {
      "spanId": "1234567893",
      "traceFlags": 1,
      "traceId": "1234567893",
      "traceState": undefined,
    },
    "_spanLimits": Object {
      "attributeCountLimit": 128,
      "attributeValueLengthLimit": Infinity,
      "eventCountLimit": 128,
      "linkCountLimit": 128,
    },
    "_spanProcessor": MultiSpanProcessor {
      "_spanProcessors": Array [
        SimpleSpanProcessor {
          "_exporter": InMemorySpanExporter {
            "_finishedSpans": [Circular],
            "_stopped": false,
          },
          "_isShutdown": false,
          "_shuttingDownPromise": Promise {},
        },
      ],
    },
    "attributes": Object {
      "tests.testSuite.duration": 3581,
      "tests.testSuite.errors": 0,
      "tests.testSuite.failed": 0,
      "tests.testSuite.name": "createTracerProvider",
      "tests.testSuite.passed": 4,
      "tests.testSuite.skipped": 0,
      "tests.testSuite.status": "PASS",
      "tests.testSuite.total": 4,
    },
    "endTime": Array [
      1642851911,
      0,
    ],
    "events": Array [],
    "instrumentationLibrary": Object {
      "name": "default",
      "version": undefined,
    },
    "kind": 0,
    "links": Array [],
    "name": "createTracerProvider",
    "parentSpanId": undefined,
    "resource": Resource {
      "attributes": Object {
        "service.name": "traceTestReportArtifact - test",
        "telemetry.sdk.language": "nodejs",
        "telemetry.sdk.name": "opentelemetry",
        "telemetry.sdk.version": "1.0.1",
      },
    },
    "startTime": Array [
      1642848330,
      0,
    ],
    "status": Object {
      "code": 0,
    },
  },
  Span {
    "_attributeValueLengthLimit": Infinity,
    "_duration": Array [
      1,
      0,
    ],
    "_ended": true,
    "_spanContext": Object {
      "spanId": "1234567899",
      "traceFlags": 1,
      "traceId": "1234567899",
      "traceState": undefined,
    },
    "_spanLimits": Object {
      "attributeCountLimit": 128,
      "attributeValueLengthLimit": Infinity,
      "eventCountLimit": 128,
      "linkCountLimit": 128,
    },
    "_spanProcessor": MultiSpanProcessor {
      "_spanProcessors": Array [
        SimpleSpanProcessor {
          "_exporter": InMemorySpanExporter {
            "_finishedSpans": [Circular],
            "_stopped": false,
          },
          "_isShutdown": false,
          "_shuttingDownPromise": Promise {},
        },
      ],
    },
    "attributes": Object {
      "tests.testCase.duration": 1,
      "tests.testCase.errors": 0,
      "tests.testCase.failed": 0,
      "tests.testCase.failure": "",
      "tests.testCase.name": "traceTestReportArtifact test junit.xml",
      "tests.testCase.passed": 0,
      "tests.testCase.skipped": 0,
      "tests.testCase.stack_trace": "",
      "tests.testCase.status": "PASS",
      "tests.testCase.total": 0,
    },
    "endTime": Array [
      1642848331,
      0,
    ],
    "events": Array [],
    "instrumentationLibrary": Object {
      "name": "default",
      "version": undefined,
    },
    "kind": 0,
    "links": Array [],
    "name": "traceTestReportArtifact test junit.xml",
    "parentSpanId": undefined,
    "resource": Resource {
      "attributes": Object {
        "service.name": "traceTestReportArtifact - test",
        "telemetry.sdk.language": "nodejs",
        "telemetry.sdk.name": "opentelemetry",
        "telemetry.sdk.version": "1.0.1",
      },
    },
    "startTime": Array [
      1642848330,
      0,
    ],
    "status": Object {
      "code": 0,
    },
  },
  Span {
    "_attributeValueLengthLimit": Infinity,
    "_duration": Array [
      178,
      0,
    ],
    "_ended": true,
    "_spanContext": Object {
      "spanId": "1234567898",
      "traceFlags": 1,
      "traceId": "1234567898",
      "traceState": undefined,
    },
    "_spanLimits": Object {
      "attributeCountLimit": 128,
      "attributeValueLengthLimit": Infinity,
      "eventCountLimit": 128,
      "linkCountLimit": 128,
    },
    "_spanProcessor": MultiSpanProcessor {
      "_spanProcessors": Array [
        SimpleSpanProcessor {
          "_exporter": InMemorySpanExporter {
            "_finishedSpans": [Circular],
            "_stopped": false,
          },
          "_isShutdown": false,
          "_shuttingDownPromise": Promise {},
        },
      ],
    },
    "attributes": Object {
      "tests.testSuite.duration": 178,
      "tests.testSuite.errors": 0,
      "tests.testSuite.failed": 0,
      "tests.testSuite.name": "traceTestReportArtifact",
      "tests.testSuite.passed": 1,
      "tests.testSuite.skipped": 0,
      "tests.testSuite.status": "PASS",
      "tests.testSuite.total": 1,
    },
    "endTime": Array [
      1642848508,
      0,
    ],
    "events": Array [],
    "instrumentationLibrary": Object {
      "name": "default",
      "version": undefined,
    },
    "kind": 0,
    "links": Array [],
    "name": "traceTestReportArtifact",
    "parentSpanId": undefined,
    "resource": Resource {
      "attributes": Object {
        "service.name": "traceTestReportArtifact - test",
        "telemetry.sdk.language": "nodejs",
        "telemetry.sdk.name": "opentelemetry",
        "telemetry.sdk.version": "1.0.1",
      },
    },
    "startTime": Array [
      1642848330,
      0,
    ],
    "status": Object {
      "code": 0,
    },
  },
  Span {
    "_attributeValueLengthLimit": Infinity,
    "_duration": Array [
      4,
      0,
    ],
    "_ended": true,
    "_spanContext": Object {
      "spanId": "1234567892",
      "traceFlags": 1,
      "traceId": "1234567892",
      "traceState": undefined,
    },
    "_spanLimits": Object {
      "attributeCountLimit": 128,
      "attributeValueLengthLimit": Infinity,
      "eventCountLimit": 128,
      "linkCountLimit": 128,
    },
    "_spanProcessor": MultiSpanProcessor {
      "_spanProcessors": Array [
        SimpleSpanProcessor {
          "_exporter": InMemorySpanExporter {
            "_finishedSpans": [Circular],
            "_stopped": false,
          },
          "_isShutdown": false,
          "_shuttingDownPromise": Promise {},
        },
      ],
    },
    "attributes": Object {
      "tests.errors": 0,
      "tests.failed": 0,
      "tests.name": "jest tests",
      "tests.passed": 5,
      "tests.retried": 0,
      "tests.skipped": 0,
      "tests.status": "PASS",
      "tests.total": 5,
      "tests.type": "junit",
    },
    "endTime": Array [
      1642848334,
      0,
    ],
    "events": Array [],
    "instrumentationLibrary": Object {
      "name": "default",
      "version": undefined,
    },
    "kind": 0,
    "links": Array [],
    "name": "test step span",
    "parentSpanId": undefined,
    "resource": Resource {
      "attributes": Object {
        "service.name": "traceTestReportArtifact - test",
        "telemetry.sdk.language": "nodejs",
        "telemetry.sdk.name": "opentelemetry",
        "telemetry.sdk.version": "1.0.1",
      },
    },
    "startTime": Array [
      1642848330,
      0,
    ],
    "status": Object {
      "code": 0,
    },
  },
]
`;
